### ISSUE TITLE
Encrypt at rest user data only accessed during live session (paranoid mode)

### ISSUE THREAD

**Issue by MrPetovan** [friendica/friendica#4841]
This would address a pet peeve of mine concerning Friendica. The project is often touted as "secure", but the threat model is unclear. Secure from whom? One thing is clear is that all the "private" posts and messages aren't private at all from anyone accessing the database, including the node admin(s).

However, there is limited room for data at rest encryption in Friendica because we provide both the server and the client, which means users can't hold a plaintext private key away from their node administrator. One solution is to encrypt symmetrically the private key using their own password and maintain the decrypted private key in session while they browse Friendica. This private key will be used to decrypt private data encrypted assymetrically using their public key before we store them in the database.

Data concerned:
- Private messages, received and sent
- Non-public posts, received and sent
- Settings?

What this method solves:
- Private data access (read/tamper) when the user is logged out (no session data)
- Casual private data access when the user is logged in (the private key has to be manually extracted from the session data then used to decrypt the private data)

What this feature **doesn't** solve:
- Private data deletion, encrypted database records still can be deleted
- Private data access reliability, if the encrypted key is lost (password lost or key deleted), then there is no way to recover it.
- Complete data access protection, there still is a window of opportunity for an indelicate admin to read and tamper a specific user's data if they monitor their activity on the node and grab the decrypted private key from the session data.

This can only be an opt-in feature as it would prevent further password recovery.

Requesting @paragonie-scott advice on cryptography/feasability/interest.

**Comment by jakimfett** [comment]
Actually, what *is* Frendica's threat model?

Having at-rest data secured would raise the threshold of effort required to compromise a user account (eg, malicious admin sniffing or an in-memory compromise of the server).

How would private-but-shared data be handled? A shared decryption key, encrypted using (each of) the user(s) public key?

**Comment by MrPetovan** [comment]
Friendica's threat model is (at least) threefold:
1. Local user threat (stolen/guessed credentials)
2. Local admin threat (private data compromission)
3. Remote user threat (man in the middle attack, message forgery)

This suggestion only addresses the local admin threat. The remote user threat is addressed by encrypting all inter-node messages even if TLS isn't used, using generated cryptographic key pairs that are stored in plaintext in the database. So, aside from strongly hashed passwords, your local admin has access to all your private data.

Does this answer your question?

**Comment by MrPetovan** [comment]
>  Can the local user threat be mitigated with multifactor authentication option(s)?

Yes, this would greatly mitigate 1. and it is planned at some point in the future.

> Is there any (conceptual-but-doable) way to limit the amount of unencrypted data available to the administrator?

This is exactly what this issue is about, with the caveats mentioned in the original post.

**Comment by Calmarius** [comment]
Sorry for the late answer I don't know if there were any step forward on this.

**The only way to solve the rogue admin problem is having end to end encryption**. **But for that you will need a stand alone client app, which is independent of the server.** This client app can generate a key pair which is used to encrypt and decrypt. So the server only ever sees encrypted payloads.

On web you cannot have end-to-end encryption because both the server and client side is controlled by the server's admin. A separate app needed which is controlled by the client.

Another advantage of a standalone app is that the server don't need to manage passwords at all. Registration involves adding a public key, and then the user can use its private key to authenticate requests.

This would solve point 2.

For solving point 1 is a matter of good key management: multiple public keys should be assigned to the account, each of them is authorized to do different things. I can imagine one key for the user's mobile phone, another key for the user's desktop PC, etc. There would be a master key which can authorize, revoke or add other keys. The private key for this master authorization key must be a very well kept secret. It should be possible to create and authenticate key authorization requests from an offline device and only submit the signed request on an online device. It's very hard to steal a private key from an offline device (especially if the medium it's stored on is a piece of paper).

**Comment by MrPetovan** [comment]
> All is a matter of good key management from the user side: multiple public keys should be assigned to the account, each of them is authorized to do different things. I can imagine one key for the user's mobile phone, another key for the user's desktop PC

Are you implying a dedicated desktop PC app as well?

**Comment by Calmarius** [comment]
That was an example. The main point of the post is separating the server and client to secure the data.
It doesn't really matter how we achieve that. It can be a mobile app or desktop app. Or can be a self hosted web client (which you can trust with your keys).

**Comment by MrPetovan** [comment]
You're absolutely right, I was just curious about what you meant when you mentioned "desktop PC".

**Comment by nathilia-peirce** [comment]
In regards of trying to increase the security of Friendica, I've mentioned before about length extension attacks that *were* possible on the session token. Ideally you should be using generated CSPRNs for session tokens, using PHP's built in `random_bytes` function (32 bytes is good). **Generate cryptographically secure pseudo-random numbers for session tokens.**

I recently took another look at the project's authentication mechanism, and it doesn't appear the form data is validated; specifically the username. Along with that, the username and password are trimmed. It's not good practice to modify the users' password. **Validate all incoming data.**

It has been a while since I went looking through the project as a whole, but MD5 and SHA1 are obsolete and vulnerable (they have been since 2010-ish). Take the time to update the hashing algorithms. I can't remember where I saw MD5 was used, but I know I saw it. You might want to consider using using an ECC algorithm to RSA, if possible. **Update cryptographic primitives.**

**And of course, don't send users generated passwords.**


## Defense in Depth

Not only can you use session tokens as a session identifier, but also as a key to encrypt session data. E.g: User logs in, and the server gives a session token in a cookie. Every time they make a request with it, the server does a look-up on the HASH of the session token (the hash of the session token is the session identifier). It's fast, and impractical to use timing attacks against this. Then the session token (NOT the session identifier) is used to decrypt the session payload. The server never stores the session token (the encryption key), but the session identifier. You can use a generic hash function (SHA256) to make the session identifier. **Encrypt the session data with the session token.**

You have suggested storing the private key in the session. Since the session data (which is on the server) is encrypted, store it in the session (NOT the cookie). When a user is making a request to post, decrypt the session data, extract the private key, encrypt the data with a new symmetric key. And then store the symmetric key encrypted with the private key.

These defense in depth strategies won't make it so your rogue admin can't decrypt the information. HOWEVER, it makes it involved. As the adversary has to wait for your session token when you make a request, and decrypt the information themselves.

This can make it harder for adversaries to obtain private information that's stored on the server, despite the fact the server is handling everything. Just remember this isn't a silver bullet. These defense in depth strategies work in a similar way as to storing hashed passwords. You're *preparing* for the event your data is breached.

**Comment by MrPetovan** [comment]
> It has been a while since I went looking through the project as a whole, but MD5 and SHA1 are obsolete and vulnerable (they have been since 2010-ish). Take the time to update the hashing algorithms. I can't remember where I saw MD5 was used, but I know I saw it. You might want to consider using using an ECC algorithm to RSA, if possible. **Update cryptographic primitives.**

Thanks for your concern, I believe you already addressed it yourself in https://github.com/friendica/friendica/pull/7725 which was a very generous contribution.

And I'm not sure I fully understand everything you said about encrypting/decrypting user's data with the session token, but I have a couple questions for you:
- How can the data be decrypted from another session? (different device, expired session)
- How can the data be encrypted on reception by the server, independently of th user's session?

Thank you for your lengthy write-up!

**Comment by nupplaphil** [comment]
I started a similar issue about encrypted session data here: #7933

**Comment by nathilia-peirce** [comment]
Each session is independent, so there isn't anyway to know which session belongs to which user, or what information each session holds. Unless you have the key. So for every login that is performed, a new session is created (hence each session is independent). With your current system, there is only one possible session per user.

Let's say we gave a client this session token (it's stored in a cookie):
```
7e7479d8e27dfd86056401000e377eb1caadaf70878225741fa9c447ed3a4d45
```

When the client makes a request to the server, the server will hash the session token:
```
session_token = "7e7479d8e27dfd86056401000e377eb1caadaf70878225741fa9c447ed3a4d45"
session_identifier = hash(session_token) // 5c074af5b4380bbcc5fedcfef382f0ab9d682338e658b936b45c92a2e5d7c9a3
```

Then the server will look in the database for the session identifier(where the question mark is the session identifier):
```
SELECT * FROM sessions WHERE identifier=?
```

If a match is found, decrypt the data:
```
key = session_token
data = decrypt(key, session_data)
```
The data in the session holds the `user_id` (to know which user they are logged in as), and the symmetric key derived from their password when they logged in (so the server can decrypt the private key).

When the user logs in, the server will store the symmetric key derived from the password in the session data (which is encrypted).

When the user tries to store sensitive information that needs encrypted, the server will use the public key of the user, to encrypt a randomly generated symmetric key, that is used to encrypt the data, and stored with the encrypted data. (this is known as hybrid-encryption schemes). **You don't need the information in sessions to encrypt data to users.**

When the user wants to read the encrypted data, they decrypt their session data with the session token, then decrypt the private key, then decrypt the symmetric key that was encrypted with the public key. Once that is done, THEN the server can finally decrypt the data with that key.

You can't/shouldn't encrypt large amounts of data with asymmetric key crypto. You encrypt the data with a symmetric key cipher, then encrypt the symmetric key with the public key.

**NOTE:** Sessions should be deleted after an inactivity of 7 to 14 days, and have a maximum life-time of 30 to 90 days. (On the server-side, because the client doesn't always delete cookies when they're supposed to)

**SECOND NOTE:** Be careful tying IPs and user agents to sessions; it can hurt usability while gaining little security improvements. People using proxies may be kicked out, even if they have a valid session token.

**Comment by MrPetovan** [comment]
Thank you so much for the follow-up, I now understand you were talking about session data and not private posts as the original issue is about. According to [this comment](https://github.com/friendica/friendica/issues/7933#issuecomment-568304407), Friendica is benefitting from the default PHP session security.

> When the user tries to store sensitive information that needs encrypted, the server will use the public key of the user, to encrypt a randomly generated symmetric key, that is used to encrypt the data, and stored with the encrypted data. (this is known as hybrid-encryption schemes). You don't need the information in sessions to encrypt data to users.
> 
> When the user wants to read the encrypted data, they decrypt their session data with the session token, then decrypt the private key, then decrypt the symmetric key that was encrypted with the public key. Once that is done, THEN the server can finally decrypt the data with that key.

I feel like an idiot for not understanding what seems to be the simplest way to explain how it is supposed to work. I'll need to read and reread this a few times, make schemas, etc.. but this doesn't have to involve you, thank you so much for your time spent writing this.

> You can't/shouldn't encrypt large amounts of data with asymmetric key crypto. You encrypt the data with a symmetric key cipher, then encrypt the symmetric key with the public key.

Thanks for the hint!

**Comment by nathilia-peirce** [comment]
@MrPetovan It's okay. Crypto is hard. But this part is specifically about the chain of encryption, and it's not exactly a small one, so it's a bit confusing to read, but easy for me to understand in my head.

**Decryption:**
1. Decrypt the session data, and extract the user's password-based encryption key.
2. Decrypt the user's private key with their password-based encryption key.
3. Decrypt the generated symmetric key with their private key.
4. Decrypt the data with the generated symmetric key.

**Encrypting data to the user:**
1. Generate a new symmetric key with `random_bytes`.
2. Encrypt the data with the generated symmetric key.
3. Encrypt the symmetric key with the public key.
4. Store the encrypted symmetric key, with the encrypted data.

You can give this symmetric key to other uses if you want them to be able to access the information also. I outlined each operation in steps, so hopefully it's more understandable than my last follow-up.

**Comment by MrPetovan** [comment]
Thank you, this is invaluable. My last question is about how to derive the symmetric encryption key you mentioned from the password besides hashing?

**Comment by nathilia-peirce** [comment]
@MrPetovan You'll probably want to use Argon2 for deriving a symmetric key. I highly recommend you use [Halite](https://github.com/paragonie/halite) to do cryptographic operations in your PHP projects. You'll need to store the salt value somewhere(probably the user row in the db), so you can generate the same encryption key every time. **If you don't store the salt, you won't be able to derive the same symmetric key, which means you won't be able to decrypt the data.**

How to [derive an encryption key in Halite](https://github.com/paragonie/halite#example-generating-a-key-from-a-password).

**Comment by MrPetovan** [comment]
We’re using libsodium since it’s been integrated to PHP 7 if I remember correctly.

**Comment by MrPetovan** [comment]
So that’s one more piece of information to store, I assume that without a salt the hashing would be easier to bruteforce from the expected password?

**Comment by nathilia-peirce** [comment]
@MrPetovan Using the salt has the same benefits, as storing hashed passwords. The symmetric keys will be different, even if other users use the same password.

**Comment by MrPetovan** [comment]
Yes, thank you again for all the knowledge. I think I will have to formalize what it would look like for Friendica because so far it's been mostly theoretical.
